### PR TITLE
CPureBlockHeader

### PR DESCRIPTION
**Pull request by micaelmalta**
Split `primitives/pureblock.(cpp|h)` from the existing `block.(cpp|h)` files, in preparation for introducing AuxPoW support.

### CODE REVIEW

**Comment by mmicael1** [comment]
@rnicoll  the CI is not running anymore...

**Comment by mmicael1** [comment]
hum `dogecoind: chainparams.cpp:122: CMainParams::CMainParams(): Assertion `consensus.hashGenesisBlock == uint256S("0x1a91e3dace36e2be3bf030a65679fe821aa1d6ef92e7c9902eb318182c355691")' failed.`

**Comment by rnicoll** [comment]
> 
> 
> hum
> 
> `dogecoind: chainparams.cpp:122: CMainParams::CMainParams(): Assertion consensus.hashGenesisBlock == uint256S("0x1a91e3dace36e2be3bf030a65679fe821aa1d6ef92e7c9902eb318182c355691")' failed.`

Guessing, but suspect it's ended up with a post-AuxPoW version number in the genesis block, or something. I'll take a look in a minute.

File: src/dogecoin.cpp [hunk 1]
```diff
@@ -6,8 +6,10 @@
 #include <boost/random/uniform_int.hpp>
 
 #include <arith_uint256.h>
+#include <chain.h>
 #include <dogecoin.h>
-#include <logging.h>
+#include <pow.h>
+#include <util/system.h>
```
Comment (rnicoll) [inline src/dogecoin.cpp:12]: Probably don't need these changes?

File: src/pow.cpp [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #include <chain.h>
 #include <dogecoin.h>
 #include <primitives/block.h>
+#include <util/system.h>
```
Comment (rnicoll) [inline src/pow.cpp:12]: Also not needed yet, I think?

File: src/protocol.cpp [hunk 3]
```diff
@@ -137,9 +137,13 @@ bool CMessageHeader::IsCommandValid() const
 
 ServiceFlags GetDesirableServiceFlags(ServiceFlags services) {
     if ((services & NODE_NETWORK_LIMITED) && g_initial_block_download_completed) {
-        return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
+        // Dogecoin: Disable segwit
+        // return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
```
Comment (rnicoll) [inline src/protocol.cpp:141]: I think this needs to be its own PR so we can revert it individually later, I'll sort that.

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -22,6 +22,7 @@
 #include <policy/policy.h>
 #include <policy/rbf.h>
 #include <primitives/transaction.h>
+#include <rpc/rawtransaction.h>
```
Comment (rnicoll) [inline src/rpc/blockchain.cpp:25]: Is this used in this PR?

File: src/test/dogecoin_tests.cpp [hunk 5]
```diff
@@ -3,6 +3,7 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <arith_uint256.h>
+#include <chain.h>
```
Comment (rnicoll) [inline src/test/dogecoin_tests.cpp:6]: Again is this needed in this PR?

**Review by rnicoll — COMMENTED** [review]
Very quick review of a few points.

**Comment by rnicoll** [comment]
Rebased #1974 on top of this (in a private branch) and sync works, so that's good!

**Comment by mmicael1** [comment]
@rnicoll all good!

**Comment by mmicael1** [comment]
NodeContext commits remove as its part of another PR

**Comment by rnicoll** [comment]
@micaelmalta Can you do a rebase of this onto 1.21-dev now, please, as we've got the Cirrus CI fixes in?

**Comment by micaelmalta** [comment]
```
==> Downloading https://dl.bintray.com/homebrew/mirror/boost_1_75_0.tar.bz2
curl: (22) The requested URL returned error: 403 Forbidden
Error: Failed to download resource "boost"
Download failed: https://dl.bintray.com/homebrew/mirror/boost_1_75_0.tar.bz2
```
👎

**Review by patricklodder — CHANGES_REQUESTED** [review]
Generic changes I'd like to see:
1. Update copyright statement for each file newly introduced, to make sure we do not erroneously assign copyright to people that didn't create those files. (i.e. Add a Dogecoin Core Developers copyright line, those can be 2014-2021)
2. Remove all code that does not touch CPureBlockHeader. I think the full set can be reverted with:

```
git checkout 1.21-dev -- src/chainparams.h
git checkout 1.21-dev -- src/init.cpp
git checkout 1.21-dev -- src/miner.h
git checkout 1.21-dev -- src/protocol.cpp
git checkout 1.21-dev -- src/qt/test/wallettests.cpp
git checkout 1.21-dev -- src/rpc/mining.cpp
git checkout 1.21-dev -- src/rpc/rawtransaction.cpp
git checkout 1.21-dev -- src/test/merkle_tests.cpp
git checkout 1.21-dev -- src/test/util/setup_common.cpp
git checkout 1.21-dev -- src/test/util/setup_common.h
git checkout 1.21-dev -- src/wallet/context.h
git checkout 1.21-dev -- src/wallet/rpcwallet.cpp
git checkout 1.21-dev -- test
git rm src/rpc/rawtransaction.h
```

I have added more comments inline.

File: src/miner.cpp [hunk 6]
```diff
@@ -127,11 +128,14 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
     assert(pindexPrev != nullptr);
     nHeight = pindexPrev->nHeight + 1;
 
-    pblock->nVersion = ComputeBlockVersion(pindexPrev, chainparams.GetConsensus());
+    const int32_t nVersion = 4;
```
[the remark this answers is not in the document]
Comment (rnicoll) [reply src/miner.cpp:131]: ComputeBlockVersion is used for BIP9. We hard-code so there's no risk of BIP9 versions being used.

[the remark this answers is not in the document]
Comment (patricklodder) [reply src/miner.cpp:131]: How does that then impact tests pre-auxpow, pre-strict dersig and pre-csv?

[the remark this answers is not in the document]
Comment (rnicoll) [reply src/miner.cpp:131]: There are test impacts, yes. I'll highlight them in a second.

**Comment by patricklodder** [comment]
This was cleaned up by @rnicoll in #2547, which has been merged. Thank you!

**Pull request closed without merging**